File: DSA/src/Questions/BinarySearch/KthElement.java
package Questions.BinarySearch;

public class KthElement {
    // Method to find the kth element in the union of two sorted arrays
    static int solve(int[] array1, int[] array2, int m, int n, int k) {
        int p1 = 0, p2 = 0, counter = 0, answer = 0;

        // Traverse both arrays
        while (p1 < m && p2 < n) {
            // If we have found k elements, break
            if (counter == k)
                break;
                // If current element of array1 is smaller, increment p1
            else if (array1[p1] < array2[p2]) {
                answer = array1[p1];
                ++p1;
            }
            // If current element of array2 is smaller, increment p2
            else {
                answer = array2[p2];
                ++p2;
            }
            // Increment the count of total elements
            ++counter;
        }
        // If we have not found k elements, check the remaining elements in the array
        // which has not been fully traversed
        if (counter != k) {
            if (p1 != m - 1)
                answer = array1[k - counter];
            else
                answer = array2[k - counter];
        }
        // Return the kth element
        return answer;
    }

    // Method to find the kth element in the union of two sorted arrays using binary
    // search
    static int BinarySearch(int[] arr1, int[] arr2, int m, int n, int k) {
        // If the size of the first array is greater than the second one, swap them
        if (m > n) {
            return BinarySearch(arr2, arr1, n, m, k);
        }

        int low = 0, high = Math.min(k, n);
        if (k > n) {
            high = m + n;
        }

        while (low <= high) {
            int cut1 = (low + high) >> 1;
            // Calculate the cut position in the second array
            int cut2 = k - cut1;
//            System.out.println("Low: " + low + " High: " + high + " Cut1: " + cut1 + " Cut2: " + cut2);
            int l1 = cut1 == 0 ? Integer.MIN_VALUE : (cut1 >= m) ? arr1[m - 1] : arr1[cut1 - 1];
            int l2 = cut2 == 0 ? Integer.MIN_VALUE : (cut2 >= n) ? arr2[n - 1] : arr2[cut2 - 1];
            int r1 = cut1 >= m ? Integer.MAX_VALUE : arr1[cut1];
            int r2 = cut2 >= n ? Integer.MAX_VALUE : arr2[cut2];
//            System.out.println("Values: L1 " + l1 + " L2 " + l2 + " R1 " + r1 + " R2 " + r2);

            // If the elements at the cut positions in both arrays are in the correct order,
            // return the maximum of them
            if (l1 <= r2 && l2 <= r1) {
                return Math.max(l1, l2);
            }
            // If the element at the cut position in the first array is greater, move
            // towards left in the first array
            else if (l1 > r2) {
                high = cut1 - 1;
            }
            // If the element at the cut position in the second array is greater, move
            // towards right in the first array
            else {
                low = cut1 + 1;
            }
        }
        // If no element is found, return -1
        return -1;
    }

    /*
     * Time Complexity : log(min(m,n)) Space Complexity : O(1)
     */
    // Main method
    public static void main(String[] args) {
        int[] array1 = {2, 3, 6, 7, 9};
        int[] array2 = {1, 4, 8, 10};
        int m = array1.length;
        int n = array2.length;
        int k = 10;
        // Find the kth element in the union of the two sorted arrays
        System.out.println(solve(array1, array2, m, n, k));
        // Find the kth element in the union of the two sorted arrays using binary
        // search
        System.out.println(BinarySearch(array1, array2, m, n, k));
    }
}

File: DSA/Python/Questions/BinarySearch/KthElement.py
"""
# Question: Kth Element of Two Sorted Arrays
# Problem Statement:
# Given two sorted arrays arr1 and arr2 of size m and n respectively, find the kth element of the merged sorted array.

# Example:
# Input: arr1 = [2,3,6,7,9], arr2 = [1,4,8,10], k = 5
# Output: 6
# Explanation: Merged array is [1,2,3,4,6,7,8,9,10], 5th element is 6
"""

from typing import List
import heapq


class KthElement:
    def find_kth_binary(self, arr1: List[int], arr2: List[int], k: int) -> int:
        """
        Binary search approach
        Time Complexity: O(log(min(m,n)))
        Space Complexity: O(1)
        """
        if len(arr1) > len(arr2):
            return self.find_kth_binary(arr2, arr1, k)

        n1, n2 = len(arr1), len(arr2)
        left, right = max(0, k - n2), min(k, n1)

        while left <= right:
            cut1 = (left + right) >> 1
            cut2 = k - cut1

            left1 = float("-inf") if cut1 == 0 else arr1[cut1 - 1]
            left2 = float("-inf") if cut2 == 0 else arr2[cut2 - 1]
            right1 = float("inf") if cut1 == n1 else arr1[cut1]
            right2 = float("inf") if cut2 == n2 else arr2[cut2]

            if left1 <= right2 and left2 <= right1:
                return max(left1, left2)
            elif left1 > right2:
                right = cut1 - 1
            else:
                left = cut1 + 1

        return -1

    def find_kth_heap(self, arr1: List[int], arr2: List[int], k: int) -> int:
        """
        Heap-based approach
        Time Complexity: O(k * log(2))
        Space Complexity: O(1)
        """
        heap = []
        i, j = 0, 0

        while i < len(arr1) and j < len(arr2) and k > 0:
            if arr1[i] < arr2[j]:
                heapq.heappush(heap, arr1[i])
                i += 1
            else:
                heapq.heappush(heap, arr2[j])
                j += 1
            k -= 1

        while i < len(arr1) and k > 0:
            heapq.heappush(heap, arr1[i])
            i += 1
            k -= 1

        while j < len(arr2) and k > 0:
            heapq.heappush(heap, arr2[j])
            j += 1
            k -= 1

        return heap[-1]

    def find_kth_merge(self, arr1: List[int], arr2: List[int], k: int) -> int:
        """
        Two-pointer merge approach
        Time Complexity: O(k)
        Space Complexity: O(1)
        """
        i, j = 0, 0
        count = 0
        result = 0

        while i < len(arr1) and j < len(arr2):
            if arr1[i] <= arr2[j]:
                result = arr1[i]
                i += 1
            else:
                result = arr2[j]
                j += 1
            count += 1
            if count == k:
                return result

        while i < len(arr1) and count < k:
            result = arr1[i]
            i += 1
            count += 1

        while j < len(arr2) and count < k:
            result = arr2[j]
            j += 1
            count += 1

        return result


def main():
    test_cases = [
        {"arr1": [2, 3, 6, 7, 9], "arr2": [1, 4, 8, 10], "k": 5, "expected": 6},
        {"arr1": [1, 3, 5, 7], "arr2": [2, 4, 6, 8], "k": 4, "expected": 4},
        {"arr1": [1, 2, 3, 4, 5], "arr2": [6, 7, 8, 9, 10], "k": 7, "expected": 7},
    ]

    solution = KthElement()
    for i, test in enumerate(test_cases):
        print(f"\nTest Case {i + 1}:")
        print(f"Array1: {test['arr1']}")
        print(f"Array2: {test['arr2']}")
        print(f"k: {test['k']}")

        result_binary = solution.find_kth_binary(test["arr1"], test["arr2"], test["k"])
        result_heap = solution.find_kth_heap(test["arr1"], test["arr2"], test["k"])
        result_merge = solution.find_kth_merge(test["arr1"], test["arr2"], test["k"])

        print(f"Binary Search Result: {result_binary}")
        print(f"Heap Result: {result_heap}")
        print(f"Merge Result: {result_merge}")
        print(f"Expected: {test['expected']}")

        assert result_binary == result_heap == result_merge == test["expected"]


if __name__ == "__main__":
    main()
